### PR TITLE
[SPARKR-161] Support reduceByKeyLocally()

### PR DESCRIPTION
**Pull request by lythesia**
I also refactored out other bykey functions, any suggestion is welcome.

### CODE REVIEW

File: pkg/R/RDD.R [hunk 1]
```diff
@@ -1442,28 +1434,79 @@ setMethod("reduceByKey",
...
             locallyReduced <- lapplyPartition(rdd, reduceVals)
             shuffled <- partitionBy(locallyReduced, numPartitions)
             lapplyPartition(shuffled, reduceVals)
           })
 
+#' Merge values by key locally
+#'
+#' This function operates on RDDs where every element is of the form list(K, V) or c(K, V).
+#' and merges the values for each key using an associative reduce function, but return the
+#' results immediately to master as R list.
```
Comment (concretevitamin) [inline pkg/R/RDD.R]: reword: "...to the driver in an R list."

File: pkg/R/RDD.R [hunk 2]
```diff
@@ -1442,28 +1434,79 @@ setMethod("reduceByKey",
...
+#' @rdname reduceByKeyLocally
+#' @seealso reduceByKey
+#' @export
+#' @examples
+#'\dontrun{
+#' sc <- sparkR.init()
+#' pairs <- list(list(1, 2), list(1.1, 3), list(1, 4))
+#' rdd <- parallelize(sc, pairs)
+#' reduced <- reduceByKeyLocally(rdd, "+")
+#' reduced[[1]] # Should be a list(1, 6)
```
Comment (concretevitamin) [inline pkg/R/RDD.R]: would be good to also show the values (or just the whole list)

File: pkg/R/RDD.R [hunk 3]
```diff
@@ -1442,28 +1434,79 @@ setMethod("reduceByKey",
...
+#' Merge values by key locally
+#'
+#' This function operates on RDDs where every element is of the form list(K, V) or c(K, V).
+#' and merges the values for each key using an associative reduce function, but return the
+#' results immediately to master as R list.
+#'
+#' @param rdd The RDD to reduce by key. Should be an RDD where each element is
+#'             list(K, V) or c(K, V).
+#' @param combineFunc The associative reduce function to use.
+#' @return An list where each element is list(K, V') where V' is the merged
```
Comment (concretevitamin) [inline pkg/R/RDD.R]: reword: "A list of elements of type list(K, V'), where V' is the reduced value for each key"

File: pkg/R/RDD.R [hunk 4]
```diff
@@ -1442,28 +1434,79 @@ setMethod("reduceByKey",
...
-                       hashVal <- as.character(hashCode(item[[1]]))
-                       if (exists(hashVal, vals)) {
-                         vals[[hashVal]] <- do.call(
-                           combineFunc, list(vals[[hashVal]], item[[2]]))
-                       } else {
-                         vals[[hashVal]] <- item[[2]]
-                         keys[[hashVal]] <- item[[1]]
-                       }
+                       item$hash <- as.character(hashCode(item[[1]]))
+                       updateOrCreatePair(item, keys, vals, pred, combineFunc, function(x) x)
```
Comment (concretevitamin) [inline pkg/R/RDD.R]: Use `identity` for `function(x) x` to avoid creating an anonymous function?

File: pkg/R/RDD.R [hunk 5]
```diff
@@ -1442,28 +1434,79 @@ setMethod("reduceByKey",
...
+          signature(rdd = "RDD", combineFunc = "ANY"),
+          function(rdd, combineFunc) {
+            reducePart <- function(part) {
+              vals <- new.env()
+              keys <- new.env()
+              pred <- function(item) exists(item$hash, keys)
+              lapply(part,
+                     function(item) {
+                       item$hash <- as.character(hashCode(item[[1]]))
+                       updateOrCreatePair(item, keys, vals, pred, combineFunc, function(x) x)
```
Comment (concretevitamin) [inline pkg/R/RDD.R]: `identity`

File: pkg/R/RDD.R [hunk 6]
```diff
@@ -1442,28 +1434,79 @@ setMethod("reduceByKey",
...
+            }
+            mergeParts <- function(accum, x) {
+              pred <- function(item) {
+                exists(item$hash, accum[[1]])
+              }
+              lapply(ls(x[[1]]),
+                     function(name) {
+                       item <- list(x[[1]][[name]], x[[2]][[name]])
+                       item$hash <- name
+                       updateOrCreatePair(item, accum[[1]], accum[[2]], pred, combineFunc, function(x) x)
```
Comment (concretevitamin) [inline pkg/R/RDD.R]: `identity`

File: pkg/R/RDD.R [hunk 7]
```diff
@@ -1513,46 +1556,29 @@ setMethod("combineByKey",
...
-                         combiners[[k]] <- do.call(mergeCombiners,
-                                                   list(combiners[[k]],
-                                                        item[[2]]))
-                       }
-                     })
-              lapply(ls(keys), function(k) {
-                      list(keys[[k]], combiners[[k]])
+                       item$hash <- as.character(item[[1]])
+                       updateOrCreatePair(item, keys, combiners, pred, mergeCombiners,
+                                          function(x) x)
```
Comment (concretevitamin) [inline pkg/R/RDD.R]: `identity`

File: pkg/R/utils.R [hunk 8]
```diff
@@ -259,3 +259,32 @@ joinTaggedList <- function(tagged_list, cnull) {
   lists <- genCompactLists(tagged_list, cnull)
   mergeCompactLists(lists[[1]], lists[[2]])
 }
+
+# Utility function to reduce a key-value list with predicate
+# Used in *ByKey functions
+# param
+#   item key-val pair
+#   keys/vals env of key/value with hashes
+#   pred predicate function
```
Comment (concretevitamin) [inline pkg/R/utils.R]: I feel like the short name `pred` doesn't convey much about what it's trying to do... what do you think about something like `updateOrCreatePred`?

File: pkg/R/utils.R [hunk 9]
```diff
@@ -259,3 +259,32 @@ joinTaggedList <- function(tagged_list, cnull) {
...
+#   update_fn update or merge function for existing pair, similar with `mergeVal` @combineByKey
+#   create_fn create function for new pair, similar with `createCombiner` @combinebykey
+updateOrCreatePair <- function(item, keys, vals, pred, update_fn, create_fn) {
```
Comment (concretevitamin) [inline pkg/R/utils.R]: nit: `updateFn` & `createFn`?

File: pkg/R/utils.R [hunk 10]
```diff
@@ -259,3 +259,32 @@ joinTaggedList <- function(tagged_list, cnull) {
...
+  # assum hashval bind to `$hash`, key/val with index 1/2
```
Comment (concretevitamin) [inline pkg/R/utils.R]: nit: `assume`

File: pkg/R/utils.R [hunk 11]
```diff
@@ -259,3 +259,32 @@ joinTaggedList <- function(tagged_list, cnull) {
...
+  hashVal <- item$hash
+  key <- item[[1]]
+  val <- item[[2]]
+  if (pred(item)) {
+    assign(hashVal, do.call(update_fn, list(get(hashVal, envir=vals), val)), envir=vals)
```
Comment (concretevitamin) [inline pkg/R/utils.R]: nit: try to follow the `arg = val` style (discussed in several of the recent PRs I think)?

File: pkg/R/utils.R [hunk 12]
```diff
@@ -259,3 +259,32 @@ joinTaggedList <- function(tagged_list, cnull) {
   lists <- genCompactLists(tagged_list, cnull)
   mergeCompactLists(lists[[1]], lists[[2]])
 }
+
+# Utility function to reduce a key-value list with predicate
+# Used in *ByKey functions
+# param
+#   item key-val pair
```
Comment (concretevitamin) [inline pkg/R/utils.R]: rename this to `pair` - I think this'd be clearer

File: pkg/R/RDD.R [hunk 13]
```diff
@@ -1382,26 +1382,18 @@ setMethod("groupByKey",
...
-                         acc <- vals[[hashVal]]
-                         acc[[length(acc) + 1]] <- item[[2]]
-                         vals[[hashVal]] <- acc
-                       } else {
-                         vals[[hashVal]] <- list(item[[2]])
-                         keys[[hashVal]] <- item[[1]]
-                       }
+                       item$hash <- as.character(hashCode(item[[1]]))
+                       updateOrCreatePair(item, keys, vals, pred,
+                                          function(vs, v) c(vs, list(v)),
```
Comment (concretevitamin) [inline pkg/R/RDD.R]: Can we emulate the old way by using [`addItemToAccumulator `](https://github.com/amplab-extras/SparkR-pkg/blob/25639cf25182a2f93e0ff7c3a76ca2844da0d29b/pkg/R/utils.R#L181)? Note that there's also `initAccumulator` which might be useful.

**Comment by concretevitamin** [comment]
Thanks @lythesia - the refactoring looks pretty neat. I did a first pass over the patch.

File: pkg/R/utils.R (hunk 12 above)
Comment (lythesia) [inline pkg/R/utils.R]: rename `item` to `pair`?

File: pkg/R/RDD.R [hunk 14]
```diff
@@ -1382,26 +1382,31 @@ setMethod("groupByKey",
...
-                         acc <- vals[[hashVal]]
-                         acc[[length(acc) + 1]] <- item[[2]]
-                         vals[[hashVal]] <- acc
-                       } else {
-                         vals[[hashVal]] <- list(item[[2]])
-                         keys[[hashVal]] <- item[[1]]
-                       }
+                       item$hash <- as.character(hashCode(item[[1]]))
+                       updateOrCreatePair(item, keys, vals, pred,
+                                          function(acc, x) {
```
Comment (concretevitamin) [inline pkg/R/RDD.R]: These two functions are in the inner most loop so to speak. Maybe pull them out of lapply? (Okay to leave them in `groupVals` I think - the number of partitions is usually not large.)

**Comment by concretevitamin** [comment]
Thanks! Merging this in.

**Pull request merged**